### PR TITLE
Slack message empty text

### PR DESCRIPTION
**Pull request by albendz**
Adding constructors to the slack message class to support more flexibility in null parameters. Additionally provided setters for use-cases where this can be added before message template is created.

https://github.com/elastic/elasticsearch/issues/30071

Replacing old pull request: https://github.com/elastic/elasticsearch/pull/31288

### CODE REVIEW

File: x-pack/plugin/watcher/src/main/java/org/elasticsearch/xpack/watcher/notification/slack/message/SlackMessage.java [hunk 1]
```diff
@@ -29,6 +29,17 @@
     final String text;
     final Attachment[] attachments;
 
+    //Note on constructors:
+    // Slack with fail to send a message if it has no attachments AND no text.
+    // Slack will succeed with attachments OR text.
+    public SlackMessage(String from, String[] to, String icon, Attachment[] attachments) {
+        this(from, to, icon, null, null); // Empty text
```
Comment (tvernum) [inline x-pack/plugin/watcher/src/main/java/org/elasticsearch/xpack/watcher/notification/slack/message/SlackMessage.java]: Shouldn't `attachments` be used here.

Comment (hub-cap) [reply x-pack/plugin/watcher/src/main/java/org/elasticsearch/xpack/watcher/notification/slack/message/SlackMessage.java]: ++, I think this is a copy-paste from the last commit just with that missing.

**Comment by hub-cap** [comment]
@elasticmachine test this please

**Comment by hub-cap** [comment]
This looks good other than that one change requested. I manually tested it against our slack, just to double check that it works. Once youve fixed that up, ill do the needful and get this merged in to master and 6.x (assuming the tests pass, which ill keep an eye on too)!

**Comment by albendz** [comment]
Thank you for reviewing; you are great with responsiveness. I'll work on getting this fixed up this week.

**Comment by hub-cap** [comment]
@albendz there was a failure in your test as well. You can reproduce it locally via the command in the wall of text thats in the build log. Note the REPRODUCE line in the message below. I typically just search for REPRODUCE if i see a build failed, in hopes that I can narrow it down to just one test instead of *all the tests* :) Thx again for the submission!!

```
14:59:20 Suite: org.elasticsearch.xpack.watcher.notification.slack.message.SlackMessageTests
14:59:20   1> [2018-06-27T16:59:20,230][INFO ][o.e.x.w.n.s.m.SlackMessageTests] [testTemplateParse]: before test
14:59:20   1> [2018-06-27T16:59:20,231][INFO ][o.e.x.w.n.s.m.SlackMessageTests] [testTemplateParse]: after test
14:59:20   1> [2018-06-27T16:59:20,234][INFO ][o.e.x.w.n.s.m.SlackMessageTests] [testUrlPathIsFiltered]: before test
14:59:20   1> [2018-06-27T16:59:20,236][INFO ][o.e.x.w.n.s.m.SlackMessageTests] [testUrlPathIsFiltered]: after test
14:59:20   1> [2018-06-27T16:59:20,239][INFO ][o.e.x.w.n.s.m.SlackMessageTests] [testTemplateParseSelfGenerated]: before test
14:59:20   1> [2018-06-27T16:59:20,239][INFO ][o.e.x.w.n.s.m.SlackMessageTests] [testTemplateParseSelfGenerated]: after test
14:59:20   1> [2018-06-27T16:59:20,243][INFO ][o.e.x.w.n.s.m.SlackMessageTests] [testToXContent]: before test
14:59:20   1> [2018-06-27T16:59:20,243][INFO ][o.e.x.w.n.s.m.SlackMessageTests] [testToXContent]: after test
14:59:20   1> [2018-06-27T16:59:20,247][INFO ][o.e.x.w.n.s.m.SlackMessageTests] [testTemplateRender]: before test
14:59:20   1> [2018-06-27T16:59:20,247][INFO ][o.e.x.w.n.s.m.SlackMessageTests] [testTemplateRender]: after test
14:59:20   1> [2018-06-27T16:59:20,249][INFO ][o.e.x.w.n.s.m.SlackMessageTests] [testCanHaveNullText]: before test
14:59:20   1> [2018-06-27T16:59:20,249][INFO ][o.e.x.w.n.s.m.SlackMessageTests] [testCanHaveNullText]: after test
14:59:20   2> REPRODUCE WITH: ./gradlew :x-pack:plugin:watcher:test -Dtests.seed=F00754B260566339 -Dtests.class=org.elasticsearch.xpack.watcher.notification.slack.message.SlackMessageTests -Dtests.method="testCanHaveNullText" -Dtests.security.manager=true -Dtests.locale=es-NI -Dtests.timezone=America/Goose_Bay
14:59:20 FAILURE 0.01s J2 | SlackMessageTests.testCanHaveNullText <<< FAILURES!
14:59:20    > Throwable #1: java.lang.AssertionError
14:59:20    > 	at __randomizedtesting.SeedInfo.seed([F00754B260566339:6906370860684020]:0)
14:59:20    > 	at org.elasticsearch.xpack.watcher.notification.slack.message.SlackMessageTests.testCanHaveNullText(SlackMessageTests.java:606)
14:59:20    > 	at java.lang.Thread.run(Thread.java:748)
14:59:20   1> [2018-06-27T16:59:20,260][INFO ][o.e.x.w.n.s.m.SlackMessageTests] [testCanHaveNullAttachments]: before test
14:59:20   1> [2018-06-27T16:59:20,260][INFO ][o.e.x.w.n.s.m.SlackMessageTests] [testCanHaveNullAttachments]: after test
```

**Comment by albendz** [comment]
I have fixed the error but clearly I didn't run the right command before I posted this. I have updated this after running targets ':x-pack:plugin:watcher:clean :x-pack:plugin:watcher:build precommit'. Let me know if I should be also running others.

**Comment by hub-cap** [comment]
So I chatted with @rjernst and he and I agreed that we probably do not need the extra constructors here. I then tested it out in our real slack integration tests and these fields are in fact already nullable fields. If you do not include these fields in the payload you submit, then they do not get sent to slack, which is what we want.

What we should do instead is just use the existing constructor (no reason to add another one for this), make both text and attachments `@Nullable`. This is from `org.elasticsearch.common` and there are other places in the codebase its used, so you can look at them. After that, you should do a check in the constructor that verifies that they are not *both* null, and if they are, throw an `IllegalArgumentException` stating they cannot both be null. I also checked this case and the slack API bombs out, which is what we are trying to prevent :)

Then of course have a test that nulls them both out and ensures the exception is thrown via `expectThrows`. 

Sorry for the churn on this, but I think this is the best solution, and it involves adding less things, so less surface area to have to test!

**Comment by albendz** [comment]
No worries! Better to vet the code before it's checked in. Summarizing what I will do:

- Remove additional constructors
- Add @Nullable to existing constructor fields for attachments and text
- Add verification that (text == null) XOR (attachments == null)
- Add test to verify both attachments and text cannot be null
- Keep added tests that verify text can be null and separately attachments can be null

**Comment by hub-cap** [comment]
sounds like a great plan!

**Comment by albendz** [comment]
Hold up on the lastest push. I have some failing tests that I will correct and forgot to run before pushing.

File: x-pack/plugin/watcher/src/test/java/org/elasticsearch/xpack/watcher/notification/slack/message/SlackMessageTests.java [hunk 2]
```diff
@@ -49,7 +49,7 @@ public void testToXContent() throws Exception {
         }
         String icon = randomBoolean() ? null : randomAlphaOfLength(10);
         String text = randomBoolean() ? null : randomAlphaOfLength(50);
-        Attachment[] attachments = randomBoolean() ? null : new Attachment[randomIntBetween(0, 2)];
+        Attachment[] attachments = !(randomBoolean() && text == null) ? null : new Attachment[randomIntBetween(0, 2)];
```
Comment (hub-cap) [inline x-pack/plugin/watcher/src/test/java/org/elasticsearch/xpack/watcher/notification/slack/message/SlackMessageTests.java]: we have a convention to not use the `!` operator in statements like this. Please use `== false` instead.

`(text != null && randomBoolean()) ? null : new .....`

This should also work

**Review by hub-cap — CHANGES_REQUESTED** [review]
One small change and then we can call this done!

**Comment by hub-cap** [comment]
@elasticmachine test this please.

File: x-pack/plugin/watcher/src/test/java/org/elasticsearch/xpack/watcher/notification/slack/message/SlackMessageTests.java (hunk 2 above)
Comment (albendz) [reply x-pack/plugin/watcher/src/test/java/org/elasticsearch/xpack/watcher/notification/slack/message/SlackMessageTests.java]: Out of curiosity, why?

I'll make this change.

**Comment by hub-cap** [comment]
@elasticmachine test this please.

**Comment by hub-cap** [comment]
@elasticmachine test this please.

I pushed some new refs to this to see if we can get CI passing!

**Review by hub-cap — APPROVED** [review]

**Comment by hub-cap** [comment]
I have merged it. Ill begin the backport process now and thank you for your contribution!

**Comment by ywelsch** [comment]
This failed in https://github.com/elastic/elasticsearch/issues/31948
@hub-cap Hold off on backporting until fixing this please

**Comment by hub-cap** [comment]
Bug fixed and backported. Ty again @albendz !!

**Comment by albendz** [comment]
I will make another attempt. Sorry about the breakage!

**Pull request merged**